binary.go

```diff
@@ -102,35 +98,59 @@ func (w *BitsWriter) Write(i interface{}) error {
 // Writes first n bytes of bs if len(bs) > n
 // Pads with padByte at the end if len(bs) < n
 func (w *BitsWriter) WriteBytesN(bs []byte, n int, padByte uint8) error {
+	if n == 0 {
+		return nil
+	}
+
 	if len(bs) >= n {
-		return w.Write(bs[:n])
+		return w.writeByteSlice(bs[:n])
 	}
 
-	if err := w.Write(bs); err != nil {
+	if err := w.writeByteSlice(bs); err != nil {
 		return err
 	}
 
 	// no bytes.Repeat here to avoid allocation
 	for i := 0; i < n-len(bs); i++ {
-		if err := w.Write(padByte); err != nil {
+		if err := w.writeFullByte(padByte); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
-func (w *BitsWriter) writeFullInt(in uint64, len int) error {
-	if w.bo == binary.BigEndian {
-		for i := len - 1; i >= 0; i-- {
-			err := w.writeFullByte(byte((in >> (i * 8)) & 0xff))
-			if err != nil {
+func (w *BitsWriter) writeByteSlice(in []byte) error {
+	// this method for push as much byte to writer in one batch
+	if len(in) == 0 {
+		return nil
+	}
+
+	if w.cacheLen != 0 {
+		for _, b := range in {
+			if err := w.writeFullByte(b); err != nil {
 				return err
 			}
 		}
+	} else {
+		if _, err := w.w.Write(in); err != nil {
```

Comment:
I'd rather create 

```go
func (w *BitsWriter) write(b []byte) error {
  if _, err := w.w.Write(b); err != nil {
    return err
  }
  if w.writeCb != nil {
    for i := range b {
      w.writeCb(b[i : i+1])
    }
  }
}
```

so that we can use it here as well as in `flushBsCache`.

What do you think?

Reply:
Yeah, i think it's good idea! I rewritten flushBsCache and writeByteSlice with this.